### PR TITLE
Job execution time limit enhancement

### LINKED ISSUE #1718: Generate wall time directive from execution timeout, etc
It would potentially be quite useful/handy/convenient to automatically generate default values for execution timeouts in tasks where directives already contain things such as wallclock times.

For example:
```cfg
[runtime]
    [[foo]]
        script = something
        ...
        [[[job submission]]]
            method = pbs
        [[[directives]]]
            -q = myqueue
            -l select=1:ncpus=6:mem=5GB:arch=linux:vntype=cray_serial
            -l walltime = 00:30:00
```
would give you a default execution timeout value of the order of PT30M should you decide to add an execution timeout handler to that task.

If you wanted to increase/decrease the value you could just define it in the task as well.

### PR DESCRIPTION
**Pull request by matthewrmshin**
Modify suite.rc spec
* New [job] section for task run time
  * execution time limit
  * execution|submission retry delays
  * execution|submission polling intervals
  * batch system name and submit command template
  * shell
* [event hooks] => [events]

From the new execution time limit setting,
* Generate time limit directives where actual directive not specified.
* Generate logic to run background or at jobs with the timeout command.
* On reaching the time limit, poll job with configurable delays.
  The default delays are 1, 2 and 7 minutes intervals,
  roughly 1, 3 and 10 minutes after reaching time limit.

Close #1718.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
@hjoliver @arjclark please review.

File: examples/inherit/single/one/suite.rc [hunk 1]
```diff
@@ -11,8 +11,8 @@ title = "User Guide [runtime] example."
                  OBS:succeed-all => bar"""
 [runtime]
     [[root]] # base namespace for all tasks (defines suite-wide defaults)
-        [[[job submission]]]
-            method = at_now
```
Comment (arjclark) [inline examples/inherit/single/one/suite.rc]: was `at_now` not different from `at`?

Comment (matthewrmshin) [inline examples/inherit/single/one/suite.rc]: This was a very old example before `at_now` was renamed to `at` (so in theory we can run `at` with different arguments, now just `now`).

File: lib/cylc/task_proxy.py [hunk 2]
```diff
@@ -1126,6 +1128,17 @@ def _prep_submit_impl(self, overrides=None):
...
+            batch_sys_conf = self._get_host_conf('batch systems')[
+                rtconfig['job']['batch system']]
+        except (TypeError, KeyError):
+            batch_sys_conf = OrderedDictWithDefaults()
+        if self.summary['execution_time_limit']:
+            # Default = 1, 2 and 7 minutes intervals, roughly 1, 3 and 10
+            # minutes after time limit exceeded
+            self.execution_time_limit_poll_try_state = (
+                TryState(delays=batch_sys_conf.get(
+                    'execution time limit polling intervals', [60, 120, 420])))
```
Comment (arjclark) [inline lib/cylc/task_proxy.py]: Did you mean 240?

Comment (arjclark) [inline lib/cylc/task_proxy.py]: Aha, missed the comment they relate to!

**Comment by arjclark** [comment]
Preliminary review and testing looks ok to me. @cylc/core - please be aware of this change (though old syntax still supported).

File: lib/cylc/cfgspec/suite.py [hunk 3]
```diff
@@ -336,11 +331,20 @@ def _coerce_final_cycletime(value, keys, args):
...
-                'command template': vdr(vtype='string'),
+            'job': {
+                'batch system': vdr(vtype='string', default='background'),
+                'batch submit command template': vdr(vtype='string'),
+                'execution polling intervals': vdr(
+                    vtype='interval_minutes_list', default=[]),
+                'execution retry delays': vdr(
+                    vtype='interval_minutes_list', default=[]),
+                'execution time limit': vdr(
+                    vtype='interval_seconds', default=[]),
```
Comment (hjoliver) [inline lib/cylc/cfgspec/suite.py]: Default shouldn't be a list?

File: doc/siterc.tex [hunk 4]
```diff
@@ -701,6 +701,28 @@ \subsection{[hosts]}
     \end{lstlisting}
 \end{myitemize}
 
+\subparagraph[{[[[[}SYSTEM{]]]]}execution time limit polling intervals]{[hosts] \textrightarrow [[HOST]] \textrightarrow [[[batch systems]]] \textrightarrow [[[[SYSTEM]]]] \textrightarrow execution time limit polling intervals}
+\label{ExecutionTimeLimitPollingIntervals}
+
+The intervals between polling after a task job (submitted to the relevant batch
+system on the relevant host) exceeds its execution time limit. The default
+setting of PT1M, PT2M, PT7M will roughly polls 1 minute, 3 minutes and 10
+minutes after the task job exceeds its execution time limit.
```
Comment (hjoliver) [inline doc/siterc.tex]: "...will roughly poll [singular] 1, 3, and 7 [not 10] minutes..."

Comment (matthewrmshin) [inline doc/siterc.tex]: (The setting is a list of intervals between polling, so the accumulated times in minutes are roughly 1, 1+2=3 and 1+2+7=10. I'll change the wording slightly to make this clearer.)

File: lib/cylc/cfgspec/suite.py (hunk 3 above)
Comment (matthewrmshin) [inline lib/cylc/cfgspec/suite.py]: I'll change this to have no default, i.e. `None`.

File: lib/cylc/task_proxy.py [hunk 5]
```diff
@@ -1259,20 +1267,40 @@ def _create_job_log_path(self):
...
             get_seconds_as_interval_string(
-                self.event_hooks['submission timeout'])
-        )
+                old_timeout - self.summary['submitted_time']))
         self.log(WARNING, msg)
         self.setup_event_handlers('submission timeout', msg)
+        return True
 
     def handle_execution_timeout(self):
         """Handle execution timeout, only called if if TASK_STATUS_RUNNING."""
```
Comment (hjoliver) [inline lib/cylc/task_proxy.py]: ("if if" - not your fault!)

File: doc/siterc.tex (hunk 4 above)
Comment (hjoliver) [inline doc/siterc.tex]: Ah, ok - fooled me, I should have read more carefully!

**Comment by matthewrmshin** [comment]
Fixed typos.

**Comment by hjoliver** [comment]
Two tests failing on my laptop VM (no proper batch systems):
```
./tests/execution-time-limit/01-at.t
./tests/execution-time-limit/00-background.t
```
with
```
ok 1 - 00-background-validate
ok 2 - 00-background-run
ok 3 - job-grep-ok
not ok 4 - job.status-grep-ok
ok 5 - job.status-grep-ok
```
It appears to be because `01/job.status` has `CYLC_JOB_EXIT=XCPU` instead of `CYLC_JOB_EXIT=ERR`.

**Comment by matthewrmshin** [comment]
The test failure is good. It means your environment is killing the job with the desired signal (XCPU). For some reason, I am unable to get XCPU in my environment. I'll modify the tests to expect either XCPU or ERR.

**Comment by matthewrmshin** [comment]
Test(s) modified.

**Comment by dpmatthews** [comment]
Please give me a chance to comment before merging - thanks

**Comment by matthewrmshin** [comment]
Branch squashed and re-based.

**Comment by hjoliver** [comment]
Review 1: this is good as far as I'm concerned, a nice improvement.

**Comment by arjclark** [comment]
@dpmatthews - am primarily waiting on your comment at this point.

**Comment by dpmatthews** [comment]
As I understand it, with this change execution timeouts become quite different.

 1. If you specify an execution time limit the execution time-out event handler will only be called if the job has not completed after the final poll (by default, 10 min after the time limit). This should only happen if the submission method you're using is not enforcing wallclock limits (unlikely) or you are unable to contact the machine to confirm the job status.

 2. If you specify an execution timeout (and not an execution time limit) then the execution time-out event handler will be called as soon as the specified time is reached. The job will also be polled to check it's latest status (possibly resulting in an update in its status and the calling of the relevant event handler). This is the old behaviour which we want users to stop using.

 3. If you specify an execution timeout AND an execution time limit then the execution timeout setting will be ignored (effectively we are deprecating execution timeout). However this is not currently documented. Note that this means it is no longer possible to trigger an event for a job which is taking longer than normal but hasn't yet reached it's wallclock limit.

I think this is fine (just want to make sure everyone is aware) but the documentation should at least state that an execution time limit over-rides an execution timeout and that execution timeout is a legacy setting which is no longer recommended.

Question: if you specify an execution time limit, does cylc detect if you have also specified the wallclock limit in the directives section? I think it's fine if it doesn't (would be difficult with pbs?) but maybe we should also document this?

Other than that I think this is a great improvement.

**Comment by matthewrmshin** [comment]
@dpmatthews I'll add your 3 points to the CUG.

> Question: if you specify an execution time limit, does cylc detect if you have also specified the wallclock limit in the directives section? I think it's fine if it doesn't (would be difficult with pbs?) but maybe we should also document this?

Looking at the logic again, it appears that I have added some limit checking. If a recognised name, (e.g. `-l walltime` for PBS or `--time` for SLURM), can be found in the `[[[directives]]]` section, then it will not add the value specified in the `execution time limit` setting to the job file. In any case, custom settings in the `[[[directives]]]` section will *always* override.

**Comment by matthewrmshin** [comment]
Branch re-based. More docs added.

**Comment by matthewrmshin** [comment]
@dpmatthews docs updated.

**Comment by dpmatthews** [comment]
I haven't reviewed all the documentation changes but the bits I wanted adding are now there - thanks

**Comment by matthewrmshin** [comment]
Branch re-based. Conflicts resolved. Tests OK (with and without global configuration) in my environment.

File: doc/cug.tex [hunk 6]
```diff
@@ -5053,6 +5061,12 @@ \subsubsection{loadleveler}
 loadleveler. This will put it back to the submitted state, until it starts
 running again.
 
+If \lstinline=execution time limit= is specified, it is used to generate the
+\lstinline=wall_clock_limit= directive. The setting is assumed to be the soft
+limit. The hard limit will be set by adding an extra minute to the soft limit.
+Do not specify the \lstinline=wall_clock_limit= directive explicitly if
+\lstinline=execution time limit= is specified, or it will cause confusion.
```
Comment (arjclark) [inline doc/cug.tex]: Functionally, what happens here? (parallel suite hat on here)

File: doc/cug.tex [hunk 7]
```diff
@@ -5171,27 +5205,32 @@ \subsubsection{slurm}
...
+\lstinline=--time= directive. Do not specify the \lstinline=--time=
+directive explicitly if \lstinline=execution time limit= is specified, or it
+will cause confusion.
+
 \subsubsection{Default Directives Provided}
 
-For job submission methods that use job file directives (PBS,
-Loadlevler, etc.) default directives are provided to set the job name
-and stdout and stderr file paths.
+For job submission methods that use job file directives (PBS, Loadlevler, etc.)
```
Comment (arjclark) [inline doc/cug.tex]: Loadleveler

File: doc/cug.tex [hunk 8]
```diff
@@ -5336,6 +5382,51 @@ \subsection{Job Killing}
...
+
+For tasks running with \lstinline=background= or \lstinline=at=, their jobs
+will be wrapped using the \lstinline=timeout= command. For all other methods,
+the relevant time limit directive will be added to their job files.
+
+The \lstinline=execution time limit= setting will also inform the suite when a
+a task job should complete by. If a task job has not reported completing within
+the specified time, the suite will poll the task job. (The default
+setting is PT1M, PT2M, PT7M. The accumulated times (in minutes) for these
+intervals will be roughly 1, 1 + 2 = 3 and 1 + 2 + 7 = 10 after a task job
```
Comment (arjclark) [inline doc/cug.tex]: + 7 = 10 *minutes*

File: doc/suiterc.tex [hunk 9]
```diff
@@ -1302,12 +1246,79 @@ \subsection{[runtime]}
     \item {\em default:} (none)
 \end{myitemize}
 
+\subparagraph[execution retry delays]{[runtime] \textrightarrow [[\_\_NAME\_\_]] \textrightarrow [[[job]]] \textrightarrow execution retry delays}
+\label{RefRetries}
+
+A list of ISO 8601 time duration/intervals after which to resubmit the task
+if it fails. The variable \lstinline=$CYLC_TASK_TRY_NUMBER= in the task
+execution environment is incremented each time, starting from 1 for the
+first try - this can be used to vary task behavior by try number.
```
Comment (arjclark) [inline doc/suiterc.tex]: *behaviour*

File: doc/suiterc.tex [hunk 10]
```diff
@@ -1302,12 +1246,79 @@ \subsection{[runtime]}
...
+\end{myitemize}
+
+\subparagraph[submission polling intervals]{[runtime] \textrightarrow [[\_\_NAME\_\_]] \textrightarrow [[[job]]] \textrightarrow submission polling intervals}
+\label{SubmissionPollingIntervals}
+
+A list of intervals, expressed as ISO 8601 duration/intervals, with optional
+multipliers, after which cylc will poll for status while the task is in the
+submitted state.
+
+For the polling task communications method this overrides the default
```
Comment (arjclark) [inline doc/suiterc.tex]: I think it should be communication~~s~~

**Comment by arjclark** [comment]
Generally looks good to me, couple of typos and a question but otherwise all good. Test battery passing in my environment.

**Comment by matthewrmshin** [comment]
All doc comments fixed.

**Comment by arjclark** [comment]
Doc changes look good to me.

**Pull request merged**